App modules: add additional app properties such as app path, immersive process, and app architecture

josephsl opened the pull request:
### Link to issue number:
Fixes #7894

### Summary of the issue:
Adds additional app properties such as path, immersive process, and architecture (WoW64).

### Description of how this pull request fixes the issue:
Adds the following properties:

* appPath: returns full executable path for the current application.
* isWindowsStoreApp: looks for a package info structure to determine if an app is a Store application (see a later comment as to why this route is used).
* appArchitecture: uses kernel32.dll's IsWoW64Process2 (Windows 10 Version 1511 and later) to return a string denoting which machien the app was written for, suitable for working with x86/x64 processes on ARM64.

### Testing performed:
Tested with various apps, both Win32 desktop apps and universal apps.

### Known issues with pull request:
None

### Change log entry:
Changes for developers:
Additional properties were added to app modules, including path for the executable (appPath), is a Windows Store app (isWindowsStoreApp), and machine architecture for the app (appArchitecture). (#7894)

josephsl committed:
App modules: add appPath property, clarify that app module packages are also supported. Re #7894.

Add appModuleHandler.AppModule.appPath property used to obtain the full path for the executable (e.g. C:\Windows\explorer.exe).
Also, in app module docstring, clarify that app module packages (appmodname/__init__.py) are also supported.

josephsl committed:
App modules: add isWindowsStoreApp property. Re #7894.

AppModule.isWindowsStoreApp will call user32.dll::IsImmersiveProcess to determine if the app is hosted inside a Windows RT container. This is useful for dealing with universla apps on Windows 10.

josephsl commented:
Hi,

I'll ask for a review once appArchitecture property is implemented and tested.

Thanks.

josephsl committed:
App modules: add appArchitecture property. Re #7894.

appModule.appArchitecture will return the architecture for which the app is written. It can be x86 (32-bit app on 32-bit Windows), AMD64 (x64 app), or ARM64 (64-bit ARM app). This is accomplished by use of kernel32.dll's IsWow64Process2 (available on Windows 10 build 10586 and later).

leonardder commented on source/appModuleHandler.py:
```diff
@@ -444,6 +454,52 @@ def _get_is64BitProcess(self):
+		These include Windows store apps on Windows 8 and 8.1, and Universal Windows Platform (UWP) apps on Windows 10.
+		@rtype: bool
+		"""
+		if winVersion.winVersion.major and winVersion.winVersion.minor == (6, 1):
```
This line is wrong. It should be:

`if (winVersion.winVersion.major, winVersion.winVersion.minor) == (6, 1):`

I prefer

`if (winVersion.winVersion.major, winVersion.winVersion.minor) < (6, 2):`

josephsl commented:
Hi, the reason for checking for equality was because we support Windows 7 SP1 and higher. I’ll edit this to make it clearer. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, August 20, 2018 1:11 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] App modules: add additional app properties such as app path, immersive process, and app architecture (#8644)

 

@leonardder commented on this pull request.

  _____  

In source/appModuleHandler.py <https://github.com/nvaccess/nvda/pull/8644#discussion_r211173688> :

> @@ -444,6 +454,52 @@ def _get_is64BitProcess(self):
               self.is64BitProcess = not res
               return self.is64BitProcess
 
+       def _get_isWindowsStoreApp(self):
+              """Whether this process is a Windows Store (immersive) process.
+              An immersive process is a Windows app that runs inside a Windows Runtime (WinRT) container.
+              These include Windows store apps on Windows 8 and 8.1, and Universal Windows Platform (UWP) apps on Windows 10.
+              @rtype: bool
+              """
+              if winVersion.winVersion.major and winVersion.winVersion.minor == (6, 1):

This line is wrong. It should be:

if (winVersion.winVersion.major, winVersion.winVersion.minor) == (6, 1):

I prefer

if (winVersion.winVersion.major, winVersion.winVersion.minor) < (6, 2):

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8644#pullrequestreview-147539126> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkFNUYi7w0sCE9Rk7dxs7epniYjqeks5uSm8CgaJpZM4WCWhu> .

josephsl committed:
App module properties: address review comments. Re #7894.

appModule.isWindowsStoreApp: check if it is below Windows 8, also fixed tuple syntax.
appModule.appArchitecture: variable renames to make it clear.

josephsl committed:
App modules/app architecture: use much improved is64bitProcess method. re #7894.

Thanks to work on ARM64 detection in appModule.is64BitProcess: use kernel32::isWow64Process2 directly and obtain either AMD64 or ARM64, falling back to AMD64 if that Windows API function isn't present (the function is present in Windows 10 Version 1511 (build 10586) and later).

francipvb commented:
What about Version fields such major, minor and revision?

josephsl commented:
Hi, not quite possible due to so many versioning schemes in use (for example, major.minor, year.major, major.minor.build, major.minor.revision, etc). Thanks. 

 

From: Francisco R. Del Roio <notifications@github.com> 
Sent: Thursday, June 13, 2019 8:01 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] App modules: add additional app properties such as app path, immersive process, and app architecture (#8644)

 

What about Version fields such major, minor and revision?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8644?email_source=notifications&email_token=AB4AXEDZL3JVVRKNX33QGGTP2MCWVA5CNFSM4FQJNBXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXVSMLQ#issuecomment-501949998> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEBRHCI7NHOX2KZF7D3P2MCWVANCNFSM4FQJNBXA> .

francipvb commented:
Hello @josephsl,

> Hi, not quite possible due to so many versioning schemes in use (for example, major.minor, year.major, major.minor.build, major.minor.revision, etc). Thanks.

Anyway, it can be nice to have such features, regardless the version scheme used by the application. Repeating code to parse the actual version of the running application for every app module that requires it is not a good way to do.

Another thing to consider is that the scheme major.minor is a standard under windows and it is relatively safe to add major and minor.

Cheers,

josephsl commented:
Hi, technically, it is possible to obtain this information. But imagine a case where a user may wish to find out NVDA’s own version via app module methods. Although it is possible to use versionInfo module for this, at first glance, due to the way version string is constructed, major.minor isn’t obvious. Also, note that what is considered major.minor could be year.major i.e. for NVDA, appModule.major will be reported as 2019 instead of 2, for instance. Not that I’m against this idea – I want to see various cases discussed before proceeding with this (by the way, I have suspended work on a PR for this due to high priority work at the moment).

 

From: Francisco R. Del Roio <notifications@github.com> 
Sent: Friday, June 14, 2019 8:06 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] App modules: add additional app properties such as app path, immersive process, and app architecture (#8644)

 

Hello @josephsl <https://github.com/josephsl> ,

Hi, not quite possible due to so many versioning schemes in use (for example, major.minor, year.major, major.minor.build, major.minor.revision, etc). Thanks.

Anyway, it can be nice to have such features, regardless the version scheme used by the application. Repeating code to parse the actual version of the running application for every app module that requires it is not a good way to do.

Another thing to consider is that the scheme major.minor is a standard under windows and it is relatively safe to add major and minor.

Cheers,

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8644?email_source=notifications&email_token=AB4AXEEISEJKM4NSKDVZUVDP2OXUFA5CNFSM4FQJNBXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXXB6WA#issuecomment-502144856> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXECSHECEGVMODII3RJ3P2OXUFANCNFSM4FQJNBXA> .

francipvb commented:
Hello,

El 14/6/2019 a las 12:16, Joseph Lee escribió:
> Hi, technically, it is possible to obtain this information. But imagine
> a case where a user may wish to find out NVDA’s own version via app
> module methods. Although it is possible to use versionInfo module for
> this, at first glance, due to the way version string is constructed,
> major.minor isn’t obvious. Also, note that what is considered
> major.minor could be year.major i.e. for NVDA, appModule.major will be
> reported as 2019 instead of 2, for instance. Not that I’m against this
> idea – I want to see various cases discussed before proceeding with this
> (by the way, I have suspended work on a PR for this due to high priority
> work at the moment).

What about naming these "first", "mid" and "last" or something like? I
think that these three parts are the most widely queried.

Probably the best approach is to implement a new lightweight class to
store these values, constructed from a plain string such productVersion
ones, replacing it by this hypothetical structure.

Sorry for the private responses, I was missing to respond to Github
notification email.

Cheers,
-- 
Cuando tus fuerzas terminan, las de mi Dios comienzan.

francipvb commented on source/appModuleHandler.py:
```diff
@@ -404,6 +404,16 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
 	chooseNVDAObjectOverlayClasses._isBase = True
 
+	def _get_appPath(self):
+		"""Returns the full path for the executable e.g. 'C:\\Windows\\explorer.exe' for Explorer.
```
Sorry but I don't see the utility of this under an appModule.

josephsl commented:
Hi,

Version class: this can be fetched easily for immersive apps as all that's required will be parsing the user mode app model structure.

Regarding first, mid, last: there are three issues with this:

1. Not exactly user friendly: although great for developers who knows what they stand for, for users, they may wish to write it as appModule.version_major instead of appModule.version_first as major/minor/build/revision terms are more prominent.
2. Does not describe schemes fully: what if the only version info expressions are first, first.last, first.mid, mid.last, etc.? At least version_first is guaranteed, but even then, there are system components that won't return a meaningful version info at all (see below).
3. No version info for processes with NULL handle: in order to fetch product info (or for that matter, anything with processes), a process handle (hProcess) must be obtained. There are situations where this fails, the most prominent example being desktop window object held by CSRSS (Client/Server Runtime Subsystem/Windows subsystem process). At least the PID can be obtained, but when trying to obtain version info, NVDA (or rather, Python) records a traceback, saying that process handle is 0 (note that process handles are relative to the calling application, not system-wide), meaning any proposal regarding parsing version info will fail for CSRSS. There are other system components that won't even show up via Windows API (although they are running), nor if they do show up somehow, hProcess will still be NULL (system process (part of ntoskrnl.exe, PID 0) will refuse to yield version info (for these, you must use sys.getwindowsversion and build versioning info). Note that we cannot catalog all sorts of system processes where hProcess will be NULL - things are changing fast.

Thanks.

josephsl commented:
Hi,

Actually, yes there is a utility for appPath: virtual machine hosts, executable path for immersive apps, and eventually, for running unit tests.

Thanks.

josephsl committed:
App architecture: rewrite the whole thing in order to support detection of 32-bit ARM apps. Re #7894.

Windows 10 Version 1903 includes foundatoins to support 32-bit ARM apps. Because this means two 32-bit architectures must be handled, rewrite app architecture method to:
1. Record architecture to arch names inside a map.
2. Call kernel32::IsWow64Process2 first and obtain arch name fromthe arch values map.
3. If IsWow64process2 fails for some reason (Windows 7 through 10 Version 1507), return appropriate architecture name based on 64-bit process status.

josephsl committed:
App architecture: mention ARM32. re #7894.

josephsl committed:
Lint fixes.

Fixes include inline comments and spaces around dictionary operator.

josephsl commented:
Hi,

There is a related issue that was discovered while working on Windows 10 Calculator app module: no product info for certain universal apps, and I'm testing a potential solution. Because that issue is different from this PR, an issue/PR pair will be created that will reference this pull request.

Thanks.

josephsl commented:
Hi,

Update on Windows Store app property: originally it asked user32::IsImmersiveProcess and answered "yes" or "no" depending on return value (1 or 0, respectively). However, this method isn't complete because:

1. What Windows reports as immersive app is actually not the case for File Explorer and other apps. In case of File Explorer, this is because it is a .Net (managed) application.
2. What Windows reports as a native (non-immersive) app turns out to be a Store app, the most notable being Office 365 package obtained from Microsoft Store.

A more accurate way is probing for app package info structure that ships with a Store package. Based on an app model XML, this structure records the following information:

* Package name (publisher.name)
* Product version (a four-part string separated by a dot).
* Architecture: the intended architecture for this app (which can be "neutral" (can run anywhere) or a processor architecture).
* Package language: the language the package is designed for (for many apps, it is either not defined (an empty string) or "neutral").
* App ID: a unique app model identifier for this package.

This structure is returned by kernel32::GetPackageFullName function as a string representation (see #10108 for more details). Every Store package (native or managed, including Store version of NVDA) will return this structure despite immersive process function returning values that do not fully reflect what the underlying package is. There might be a use case for IsImmersiveProcess - perhaps looking to see if we're dealing with a managed application.

Note that the above remark is irrelevant on Windows 7 as it does not come with package info functions nor immersive process checker routine.

Thanks.

josephsl committed:
App properties: use a dedicated (private) method for fetching package info for Microsoft Store apps. Re #7894.

User32.dll::IsImmersiveProcess won't quite cut it for Store apps:
* File Explorer is seen as an immersive app but it is not really a Store application.
* Converted desktop apps are seen as not immersive but they are Store apps.
The more accurate way to determine this information is looking for app package info, which can be done by calling kernel32.dll::GetPackageFullName. This returns a string representation of app name, version, architecture, language, and app ID (the first two are used in #4259). Every Store app comes with app model XML file that provides data for this structure (Store version of NVDA is no exception). Thus use this method, which is now part of a private function in the base app module so not only product name and version setter can use it, but also Store app property can call this method.

josephsl committed:
Product name and version for Store apps: use the dedicated package info function. Re #7894.

Use the dedicated package info function for obtaining product name and version:
* If the package info is a string, info will be parsed.
* If it is None, executable file info will be fetched.

josephsl committed:
Windows Store property: use package info method to determine if the underlying process is a Windows Store app. Re #7894.

User32.dll::IsImmersiveProcess won't work all the way, but package info fetching will (on Windows 8 and later). Therefore use the package info method to determine if a given process is a Windows Store app.

josephsl committed:
Windows Store app property: use platform version tuple from sys.getwindowsversion(). Re #7894.

josephsl commented:
Hi,

Suspended until further notice in light of #11006. 

Thanks.

Adriani90 commented:
@josephsl I understand #11006 addresses new pull requests, not already existing pull requests. If we don't continue work on existing pull requests, then we run into the same issue later on.
I suggest if possible, continue work on existing pull requests, so that the backlog can be reviewed and finalized as soon as possible.

josephsl commented:
Hi, the reason for suspending it for now is to let higher priority PR’s to get a chance to get reviewed. Although this PR is important, it isn’t a high priority work. Thanks.

 

From: Adriani90 <notifications@github.com> 
Sent: Tuesday, April 14, 2020 1:59 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] App modules: add additional app properties such as app path, immersive process, and app architecture (#8644)

 

@josephsl <https://github.com/josephsl>  I understand #11006 <https://github.com/nvaccess/nvda/issues/11006>  addresses new pull requests, not already existing pull requests. If we don't continue work on existing pull requests, then we run into the same issue later on.
I suggest if possible, continue work on existing pull requests, so that the backlog can be reviewed and finalized as soon as possible.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8644#issuecomment-613677994> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEBI2FJ6GUGPDN7VTQ3RMTE7ZANCNFSM4FQJNBXA> .

josephsl commented:
Hi,

The PR is trivial enough for a fast forward review, but I imagine there might be scenarios where testing might be required. Even then, I think there are PR's that are more urgent than this one (depends on PR tags).

Thanks.

feerrenrut commented:
> Suspended until further notice in light of #11006.

Please continue to work on this PR. #11006 requests no new PR's. This PR already exists and is thus in the backlog of PR's that we want to address before accepting new ones. We are trying to stop PR's from going stale, and then becoming hard to merge. 

Forgive me for replying to you multiple times, I want to avoid the community being mislead.

josephsl commented:
Hi, rebasing was successful, so it is ready for review again. Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Wednesday, April 15, 2020 3:53 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] App modules: add additional app properties such as app path, immersive process, and app architecture (#8644)

 

Suspended until further notice in light of #11006 <https://github.com/nvaccess/nvda/issues/11006> .

Please continue to work on this PR. #11006 <https://github.com/nvaccess/nvda/issues/11006>  requests no new PR's. This PR already exists and is thus in the backlog of PR's that we want to address before accepting new ones. We are trying to stop PR's from going stale, and then becoming hard to merge.

Forgive me for replying to you multiple times, I want to avoid the community being mislead.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8644#issuecomment-613967677> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEFLQBJDLBVM4K2JT3LRMWGZZANCNFSM4FQJNBXA> .

michaelDCurran commented on source/appModuleHandler.py:
```diff
@@ -362,6 +363,22 @@ def _getExecutableFileInfo(self):
+		# Others such as Store version of Office are not truly hosted apps but are distributed via Store.
+		length = ctypes.c_uint()
+		buf = ctypes.windll.kernel32.GetPackageFullName(self.processHandle, ctypes.byref(length), None)
+		packageFullName = ctypes.create_unicode_buffer(buf)
```
The argument to create_unicode_buffer should be length, not buf I believe? 
According to https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getpackagefullname
GetPackageFullname returns ERROR_SUCCESS or an error code. The needed length is written to the length parameter.

josephsl committed:
Merge branch 'master' into i7894appModProperties

josephsl commented:
Hi, you’re right. Thanks for clarifying this.

 

From: Michael Curran <notifications@github.com> 
Sent: Sunday, July 5, 2020 7:31 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] App modules: add additional app properties such as app path, immersive process, and app architecture (#8644)

 

@michaelDCurran requested changes on this pull request.

  _____  

In source/appModuleHandler.py <https://github.com/nvaccess/nvda/pull/8644#discussion_r449954156> :

> @@ -362,6 +363,22 @@ def _getExecutableFileInfo(self):
               fileinfo = getFileVersionInfo(fileName, "ProductName", "ProductVersion")
               return (fileinfo["ProductName"], fileinfo["ProductVersion"])
 
+       def _getImmersivePackageInfo(self):
+              # Used to obtain full package structure for a hosted app.
+              # The package structure consists of product name, version, architecture, language, and app ID.
+              # This is useful for confirming whether an app is hosted or not despite an app reporting otherwise.
+              # Some apps such as File Explorer says it is an immersive process but error 15700 is shown.
+              # Others such as Store version of Office are not truly hosted apps but are distributed via Store.
+              length = ctypes.c_uint()
+              buf = ctypes.windll.kernel32.GetPackageFullName(self.processHandle, ctypes.byref(length), None)
+              packageFullName = ctypes.create_unicode_buffer(buf)

The argument to create_unicode_buffer should be length, not buf I believe?
According to https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getpackagefullname
GetPackageFullname returns ERROR_SUCCESS or an error code. The needed length is written to the length parameter.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8644#pullrequestreview-442740272> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXECK4IONEDOVJRDFZI3R2EZONANCNFSM4FQJNBXA> .

josephsl committed:
App modules/immersive package info: use length instead of unused buffer to obtain actual package info length (addressing review action). Re #7894.
